staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go

```diff
@@ -594,10 +600,28 @@ func patchResource(
 				}
 				originalObjJS, originalPatchedObjJS = originalJS, patchedJS
 			case types.StrategicMergePatchType:
-				originalMap, patchMap, err := strategicPatchObject(codec, currentObject, patchJS, objToUpdate, versionedObj)
+				// Since the patch is applied on versioned objects, we need to convert the
+				// current object to versioned representation first.
+				gvks, _, err := typer.ObjectKinds(versionedObj)
 				if err != nil {
 					return nil, err
 				}
+				gvk := gvks[0]
+				currentVersionedObject, err := convertor.ConvertToVersion(currentObject, gvk.GroupVersion())
+				if err != nil {
+					return nil, err
+				}
+				versionedObjToUpdate, err := creater.New(gvk)
+				if err != nil {
+					return nil, err
+				}
+				originalMap, patchMap, err := strategicPatchObject(codec, currentVersionedObject, patchJS, versionedObjToUpdate, versionedObj)
```

Comment:
OK - so I managed to changed it to use unsafe conversions, which should pretty much be the same as before (when serializing/deserializing to json).